onnxruntime/core/framework/session_state.cc

```diff
@@ -1196,6 +1203,17 @@ static Status VerifyEachNodeIsAssignedToAnEp(const Graph& graph, const logging::
   }
 #endif  // !defined(ORT_MINIMAL_BUILD)
 
+  // Silent fallback from GPU/NPU to CPU nodes can cause performance issues due to memory copies and frequent stalls.
+  // If the user explicitly included the CPU provider anyway, then remain silent, but if it was implicitly added,
+  // and unexpected fallback happened to a non-preferred provider, warn the user.
+  size_t explicit_provider_count = providers.NumProviders() - providers.GetCpuProviderWasImplicitlyAdded() ? 1 : 0;
+  if (node_placement_set.size() > explicit_provider_count) {
```

Comment:
Duh. 😑

Reply:
Ah, now I remember why. The `node_placements` was under the `#if !defined(ORT_MINIMAL_BUILD)` and so it would be completely unpopulated in minimal builds (and thus not give a warning message). What are the guiding principles for ORT_MINIMAL_BUILD when deciding what to include/exclude? I kinda wanted the message to exist on minimal builds, just clarifying to rerun it with verbose on a non-minimal build.

Reply:
The current setup uses hashes in the session state finalization to finish node assignments so iirc the info wasn't available (vs. all nodes being assigned during partitioning). With Ed's new setup it would be, however we try and avoid adding anything optional to the base minimal build. Given the user has to first convert an onnx model to ort format before using a minimal build, they should see warnings about node placement earlier. Is that insufficient?

Reply:
@skottmckay That seems reasonable to me. I'll move it under the `ORT_MINIMAL_BUILD` and delete the `node_placement_set`.

Reply:
> Is this a DML specific requirement that all nodes are assigned to that EP? Is so the DML EP can enforce that in GetCapability (and this behavior should be configured via the DML EP).

@skottmckay It's not specific to DML, as it's applicable to other NPU/GPU EP's due to extra synchronization and memory copies between devices. DML EP functions with nodes falling back to CPU (if we didn't there would be models with newer opsets that wouldn't even execute, because the DML EP is a bit behind CPU implementation), but it's a perf surprise to many customers. Ideally you want partitions to be larger with little fragmentation, and you want to set free dimensions so that constant folding can eliminate most of those Shape ops and slices of shapes (which benefits CUDA too).

@jgauth I'm curious, for your case, are you using the DML EP or another?

Reply:
@skottmckay Wait, updating `node_placements` is under `is_verbose`. So now nobody will see this message unless `is_verbose` is also on, which defeats the point of visibility in the norm case. 🤔

Reply:
In that case keeping node_placement_set should be fine. Just rename it to node_placement_provider_set to reflect its actual contents. Compared to node_placements (which contains the names of all nodes), node_placement_provider_set seems to be a good option (just a handful of providers).

Reply:
Renamed. Pushed.